### PR TITLE
Improve builds runner

### PR DESCRIPTION
**Pull request by MaksimZhukov**
In scope of this PR we improve builds runner. We've added 
- retry logic for builds queuing;
- `WAIT_FOR_BUILDS` flag in order to manage waiting of queued builds results;
- opportunity to specify link to the `versions-manifest.json` instead of tool versions list in order to rebuild all uploaded versions

image

### CODE REVIEW

File: azure-devops/run-ci-builds.ps1 [hunk 1]
```diff
@@ -3,31 +3,80 @@ param (
...
+        [string] $ToolVersions
+    )
+    
+    [string[]] $versionsList = @()
+    if ($ToolVersions){
+        $versionsList = $ToolVersions.Split(',')
+    } else {
+        Write-Host "Get the list of releases from $ManifestLink"
+        $releases = Invoke-RestMethod $ManifestLink -MaximumRetryCount $Retries -RetryIntervalSec $RetryIntervalSec
+        $releases | ForEach-Object { $versionsList += $_.version }
```
Comment (al-cheb) [inline azure-devops/run-ci-builds.ps1:32]: nitpick: `$versionsList += $releases.version`

File: azure-devops/run-ci-builds.ps1 [hunk 2]
```diff
@@ -3,31 +3,80 @@ param (
...
+    [UInt32] $RetryIntervalSec = 30,
+    [UInt32] $RetryCount = 20
 )
 
 Import-Module (Join-Path $PSScriptRoot "azure-devops-api.ps1")
 Import-Module (Join-Path $PSScriptRoot "build-info.ps1")
 
+function Get-ToolVersions {
+    param (
+        [Parameter(Mandatory)] [string] $ManifestLink,
```
Comment (al-cheb) [inline azure-devops/run-ci-builds.ps1:20]: Difficult to understand why should we define `ManifestLink,RetryIntervalSec,Retries` parameters when we don't use them at all when the variable ToolVersions is defined?

File: azure-devops/run-ci-builds.ps1 [hunk 3]
```diff
@@ -3,31 +3,80 @@ param (
...
+        [Parameter(Mandatory)] [UInt32] $RetryIntervalSec,
+        [Parameter(Mandatory)] [UInt32] $Retries,
+        [string] $ToolVersions
+    )
+    
+    [string[]] $versionsList = @()
+    if ($ToolVersions){
```
Comment (vmapetr) [inline azure-devops/run-ci-builds.ps1:27]: Nitpick: missing space between condition and opening brace.

Comment (MaksimZhukov) [reply azure-devops/run-ci-builds.ps1]: Thanks!

File: azure-devops/run-ci-builds.ps1 (hunk 1 above)
Comment (MaksimZhukov) [reply azure-devops/run-ci-builds.ps1]: Good point, thanks!

File: azure-devops/run-ci-builds.ps1 (hunk 2 above)
Comment (maxim-lobanov) [reply azure-devops/run-ci-builds.ps1:20]: Agree, can we do 2 separate functions?

File: azure-devops/run-ci-builds.ps1 [hunk 4]
```diff
@@ -3,31 +3,80 @@ param (
...
+        [string] $ToolVersions
+    )
+    
+    [string[]] $versionsList = @()
+    if ($ToolVersions) {
+        $versionsList = $ToolVersions.Split(',')
+    } else {
+        Write-Host "Get the list of releases from $ManifestLink"
+        $releases = Invoke-RestMethod $ManifestLink -MaximumRetryCount $Retries -RetryIntervalSec $RetryIntervalSec
+        $versionsList = $releases.version
```
Comment (maxim-lobanov) [inline azure-devops/run-ci-builds.ps1:32]: Do we want to log versions that we have received?

File: azure-devops/run-ci-builds.ps1 [hunk 5]
```diff
@@ -3,31 +3,80 @@ param (
...
 )
 
 Import-Module (Join-Path $PSScriptRoot "azure-devops-api.ps1")
 Import-Module (Join-Path $PSScriptRoot "build-info.ps1")
 
+function Get-ToolVersions {
+    param (
+        [Parameter(Mandatory)] [string] $ManifestLink,
+        [Parameter(Mandatory)] [UInt32] $RetryIntervalSec,
+        [Parameter(Mandatory)] [UInt32] $Retries,
```
Comment (maxim-lobanov) [inline azure-devops/run-ci-builds.ps1:22]: Just use global params, don't need to pass them

File: azure-devops/run-ci-builds.ps1 [hunk 6]
```diff
@@ -3,31 +3,80 @@ param (
...
+                Write-Host "There is an error during build starting:`n $_"
+                $Retries--
+            
+                if ($Retries -eq 0)
+                {
+                    Write-Host "Build can't be queued, please try later"
+                    exit 1
+                }
+            
+                Write-Host "Waiting 30 seconds before retrying. Retries left: $Retries"
```
Comment (maxim-lobanov) [inline azure-devops/run-ci-builds.ps1:75]: It is too difficult for read the code.
Could you please move retry logic to helper function like `Invoke-RestMethodWithRetry`

Comment (maxim-lobanov) [reply azure-devops/run-ci-builds.ps1:75]: Can we also add retry logic to Wait-Builds method?

File: azure-devops/run-ci-builds.ps1 (hunk 2 above)
Comment (MaksimZhukov) [reply azure-devops/run-ci-builds.ps1:20]: We discussed with @maxim-lobanov and decided not to make changes

File: azure-devops/run-ci-builds.ps1 (hunk 4 above)
Comment (MaksimZhukov) [reply azure-devops/run-ci-builds.ps1:32]: Yes, thanks

File: azure-devops/run-ci-builds.ps1 (hunk 5 above)
Comment (MaksimZhukov) [reply azure-devops/run-ci-builds.ps1:22]: We discussed with @maxim-lobanov and decided not to make changes

File: azure-devops/run-ci-builds.ps1 (hunk 6 above)
Comment (MaksimZhukov) [reply azure-devops/run-ci-builds.ps1]: Good point, thanks! I've moved retry logic to the `AzureDevOpsApi` class

**Pull request merged**